src/main/java/hudson/tasks/test/AbstractTestResultAction.java

```diff
@@ -287,29 +350,249 @@ public TestResult findCorrespondingResult(String id) {
         return Collections.emptyList();
     }
 
+    /**
+     * Start for Test Result Trends.
+     * From here onwards start the code area responsible for generating various test result trends.
+     */
+
+    public hudson.tasks.junit.TestResult loadXml() {
+        return null;
+    }
+
+    /**
+     * A method for getting the list of packages for all levels of hierarchy.
+     *
+     * @return Array of packages of all hierarchies.
+     */
+    public final String[] getProjectList() {
+
+        /*
+         * If project list is already ready no need to construct again.
+         */
+        if (null != projectList)
+            return projectList;
+        hudson.tasks.junit.TestResult r = loadXml();
+        Collection<SuiteResult> suiteList = r.getSuites();
+
+        /*
+         * A set for the package names.
+         */
+        Map<Integer, HashSet<String>> projectSet = new HashMap<>();
+        int projectCount = 0;
+        final int projectCountLimit = 50;
+        int levelCap = Integer.MAX_VALUE;
+        for (SuiteResult suiteResult : suiteList) {
+            String suiteName = suiteResult.getName();
```

Comment:
We should try to avoid duplication like we discussed e.g
```
org.salesforce.apache -1 
org => 
org.salesforce => 
org.salesforce.apache => 
No other org.salesforce
```